File: day 12/AOC12.py
from collections import deque

def parse_map(input_map):
    return [list(row) for row in input_map.splitlines()]

def get_neighbors(x, y, rows, cols):
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    for dx, dy in directions:
        nx, ny = x + dx, y + dy
        if 0 <= nx < rows and 0 <= ny < cols:
            yield nx, ny

def flood_fill(grid, x, y, visited):
    rows, cols = len(grid), len(grid[0])
    queue = deque([(x, y)])
    region_type = grid[x][y]
    region_cells = []
    visited[x][y] = True

    while queue:
        cx, cy = queue.popleft()
        region_cells.append((cx, cy))

        for nx, ny in get_neighbors(cx, cy, rows, cols):
            if not visited[nx][ny] and grid[nx][ny] == region_type:
                visited[nx][ny] = True
                queue.append((nx, ny))

    return region_cells

def calculate_perimeter(grid, region_cells):
    perimeter = 0
    rows, cols = len(grid), len(grid[0])

    for x, y in region_cells:
        for nx, ny in get_neighbors(x, y, rows, cols):
            # Increment perimeter if the neighbor is not part of the same region
            if grid[nx][ny] != grid[x][y]:
                perimeter += 1
        # Add edge contributions explicitly
        if x == 0:
            perimeter += 1
        if x == rows - 1:
            perimeter += 1
        if y == 0:
            perimeter += 1
        if y == cols - 1:
            perimeter += 1

    return perimeter


def calculate_total_price(input_map):
    grid = parse_map(input_map)
    rows, cols = len(grid), len(grid[0])
    visited = [[False] * cols for _ in range(rows)]
    total_price = 0

    for x in range(rows):
        for y in range(cols):
            if not visited[x][y]:
                region_cells = flood_fill(grid, x, y, visited)
                area = len(region_cells)
                perimeter = calculate_perimeter(grid, region_cells)
                price = area * perimeter
                total_price += price

    return total_price

# Read input from file
with open("perem.txt", "r") as file:
    input_map = file.read()

print(calculate_total_price(input_map))

File: day 12/AOC12.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Pos struct {
	x, y int
}

func parseMap(inputMap string) [][]rune {
	lines := strings.Split(strings.TrimSpace(inputMap), "\n")
	grid := make([][]rune, len(lines))
	for i, line := range lines {
		grid[i] = []rune(line)
	}
	return grid
}

func getNeighbors(x, y, rows, cols int) []Pos {
	directions := []Pos{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
	var neighbors []Pos
	for _, dir := range directions {
		nx, ny := x+dir.x, y+dir.y
		if nx >= 0 && nx < rows && ny >= 0 && ny < cols {
			neighbors = append(neighbors, Pos{nx, ny})
		}
	}
	return neighbors
}

func floodFill(grid [][]rune, x, y int, visited [][]bool) []Pos {
	rows, cols := len(grid), len(grid[0])
	queue := []Pos{{x, y}}
	regionType := grid[x][y]
	var regionCells []Pos
	visited[x][y] = true

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		regionCells = append(regionCells, curr)

		for _, neighbor := range getNeighbors(curr.x, curr.y, rows, cols) {
			if !visited[neighbor.x][neighbor.y] && grid[neighbor.x][neighbor.y] == regionType {
				visited[neighbor.x][neighbor.y] = true
				queue = append(queue, neighbor)
			}
		}
	}

	return regionCells
}

func calculatePerimeter(grid [][]rune, regionCells []Pos) int {
	rows, cols := len(grid), len(grid[0])
	perimeter := 0

	for _, cell := range regionCells {
		for _, neighbor := range getNeighbors(cell.x, cell.y, rows, cols) {
			if grid[neighbor.x][neighbor.y] != grid[cell.x][cell.y] {
				perimeter++
			}
		}
		// Add edge contributions
		if cell.x == 0 {
			perimeter++
		}
		if cell.x == rows-1 {
			perimeter++
		}
		if cell.y == 0 {
			perimeter++
		}
		if cell.y == cols-1 {
			perimeter++
		}
	}

	return perimeter
}

func calculateTotalPrice(inputMap string) int {
	grid := parseMap(inputMap)
	rows, cols := len(grid), len(grid[0])
	visited := make([][]bool, rows)
	for i := range visited {
		visited[i] = make([]bool, cols)
	}

	totalPrice := 0
	for x := 0; x < rows; x++ {
		for y := 0; y < cols; y++ {
			if !visited[x][y] {
				regionCells := floodFill(grid, x, y, visited)
				area := len(regionCells)
				perimeter := calculatePerimeter(grid, regionCells)
				price := area * perimeter
				totalPrice += price
			}
		}
	}

	return totalPrice
}

func main() {
	content, err := os.ReadFile("perem.txt")
	if err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		return
	}

	result := calculateTotalPrice(string(content))
	fmt.Println(result)
}
